Support device probing and permission request on Android

wuwbobo2021 opened the pull request:
Closes #86.

`nusb::list_buses` (for rooted usage) is disabled because of some problems caused by `port_chain`, `device_version`, `max_packet_size_0` unavailable on Android. Could they be changed to `Option`s?

I added `#![allow(dead_code)]` in `platform/linux_usbfs/mod.rs` because of the conditional compilation condition for Android caused many dead code warnings, maybe it can be solved in a better way.

Sorry for being a bit sloppy, here's how I did the initial test:

Please make sure the JDK, Android SDK and NDK is installed and configured, and the Rust target `aarch64-linux-android` is installed. Install `cargo-apk` and make sure the release keystore is configured.

Save the code files posted below, run `cargo apk build -r`; connect the Android phone to the PC via USB, then configure the adbd TCP port (`adb tcpip 5555`) and connect to it (`adb connect xxx.xxx.xxx.xxx:5555`). Install the APK package with `adb install -r <apk_file>` (the .apk file is located in `target/release/apk`).

Run `adb logcat android_nusb_test:D '*:S'` On PC for tracing, start the installed "android_nusb_test" on the phone; try to connect and disconnect some USB devices via an OTG USB converter.

(this draft code for testing can be changed in many ways in order to test all features thoroughly)

<details>
<summary>Cargo.toml</summary>

```toml
[package]
name = "android-nusb-test"
version = "0.1.0"
edition = "2024"
publish = false

[dependencies]
log = "0.4"
futures-lite = "2.0"
android-activity = { version = "0.6", features = ["native-activity"] }
android_logger = "0.14"
nusb = { path = ".." }

[lib]
name = "android_nusb_test"
crate-type = ["cdylib"]
path = "lib.rs"

[package.metadata.android]
package = "com.example.android_nusb_test"
build_targets = ["aarch64-linux-android"]
resources = "./res"

[package.metadata.android.sdk]
min_sdk_version = 16
target_sdk_version = 30

[[package.metadata.android.uses_feature]]
name = "android.hardware.usb.host"
required = true

[[package.metadata.android.application.activity.intent_filter]]
actions = ["android.hardware.usb.action.USB_DEVICE_ATTACHED"]

# Please check <https://github.com/rust-mobile/cargo-apk/pull/67> if it fails.
# Otherwise comment out the lines below (request for permission purely at runtime).
[[package.metadata.android.application.activity.meta_data]]
name = "android.hardware.usb.action.USB_DEVICE_ATTACHED"
resource = "@xml/device_filter"
```

</details>

<details>
<summary>lib.rs</summary>

```rust
use android_activity::{AndroidApp, MainEvent, PollEvent};
use log::info;
use nusb::MaybeFuture;
use std::time::Duration;

#[unsafe(no_mangle)]
fn android_main(app: AndroidApp) {
    android_logger::init_once(
        android_logger::Config::default()
            .with_max_level(log::LevelFilter::Info)
            .with_tag("android_nusb_test"),
    );

    if let Some(startup_dev) = nusb::check_startup_intent() {
        info!("*** Startup intent: {startup_dev:#?}");
        assert!(startup_dev.open().wait().is_ok());
    } else {
        info!("Listing connected devices...");
        for dev in nusb::list_devices().wait().unwrap() {
            info!("*** {dev:#?}");
        }
    }

    // XXX: tell this thread to stop on activity's stop or destroy event
    let _back_thread = std::thread::spawn(|| hotplug_watch());

    let mut on_destroy = false;
    loop {
        app.poll_events(
            Some(Duration::from_secs(1)), // timeout
            |event| match event {
                PollEvent::Main(MainEvent::Start) => {
                    info!("Main Start.");
                }
                PollEvent::Main(MainEvent::Resume { loader: _, .. }) => {
                    info!("Main Resume.");
                }
                PollEvent::Main(MainEvent::Pause) => {
                    info!("Main Pause.");
                }
                PollEvent::Main(MainEvent::Stop) => {
                    info!("Main Stop.");
                }
                PollEvent::Main(MainEvent::Destroy) => {
                    info!("Main Destroy.");
                    on_destroy = true;
                }
                _ => (),
            },
        );
        if on_destroy {
            return;
        }
    }
}

fn hotplug_watch() {
    info!("Testing hotplug watch...");
    let watcher = futures_lite::stream::block_on(nusb::watch_devices().unwrap());
    for event in watcher {
        println!("{:#?}", event);
        match event {
            nusb::hotplug::HotplugEvent::Connected(dev) => {
                if !dev.has_permission().unwrap() {
                    info!("*** {dev:#?}");
                    let perm_req = dev.request_permission().unwrap_or(None);
                    if let Some(req) = perm_req {
                        info!("Performing permission request...");
                        let result = futures_lite::future::block_on(req);
                        info!("Request result: {result:?}");
                        if !result {
                            continue;
                        }
                    }
                }
                let conn = dev.open().wait().unwrap();
                info!("*** Opened, printing USB configurations:");
                for conf in conn.configurations() {
                    info!("{:#?}", conf);
                }
                std::thread::sleep(Duration::from_secs(1));
                info!("Closing the device.");
            }
            nusb::hotplug::HotplugEvent::Disconnected(dev_id) => {
                info!("*** Device disconnected: ({dev_id:?}).");
                continue;
            }
        }
    }
}
```

</details>

`res/xml/device_filter.xml`:

```
<?xml version="1.0" encoding="utf-8"?>

<resources>
    <usb-device />
</resources>
```

(this means no filtering, any device plugged into the phone can trigger the prompt asking whether or not to open this "app".)

wuwbobo2021 commented:
The sloppy multi-thread `open()` test is inspired by <https://github.com/jni-rs/jni-rs/issues/574>. It's about the safety of the only unsafe operation in `android.rs` (`OwnedFd::from_raw_fd`). My result shows it's okay on my device, but the access of `usb_manager` singleton might be changed to be protected by `Mutex` to make sure of compatability.

Another test shows that getting the Java `UsbManager` in different threads via `getSystemService` gets the same `UsbManager` instance:

<details>

<summary>Click to expand</summary>

```rust
    // insert in `usb_manager()`
    use std::sync::mpsc;
    let usb_man_2 = usb_man.clone();
    let (tx, rx) = mpsc::channel();
    std::thread::spawn(move || {
        jni_with_env(|env| {
            let context = android_context();
            let usb_service_id = USB_SERVICE.new_jobject(env)?;
            let usb_man = env
                .call_method(
                    context,
                    "getSystemService",
                    "(Ljava/lang/String;)Ljava/lang/Object;",
                    &[(&usb_service_id).into()],
                )
                .get_object(env)?;
            tx.send(usb_man.is_same_object(&usb_man_2, env)).unwrap();
            Ok(())
        }).unwrap();
    });
    assert!(rx.recv().unwrap());
```

</details>

Yet another test shows that probing the same device in different threads gets different Java `UsbDevice` (info) instances (kind of senseless):

<details>

<summary>Click to expand</summary>

```rust
            // insert in `list_devices()`
            if !devices.is_empty() {
                use std::sync::mpsc;
                let (tx, rx) = mpsc::channel();
                std::thread::spawn(move || {
                    jni_with_env(|env| {
                        let ref_dev_list = env
                            .call_method(usb_man, "getDeviceList", "()Ljava/util/HashMap;", &[])
                            .get_object(env)?;
                        let map_dev = env.get_map(&ref_dev_list)?;
                        let mut iter_dev = map_dev.iter(env)?;
                        if let Some((_, dev)) = iter_dev.next(env)? {
                            tx.send(build_device_info(env, &dev)?).unwrap();
                        }
                        Ok(())
                    }).unwrap();
                });
                assert!(!rx.recv().unwrap().devinst.is_same_object(devices[0].devinst.as_ref(), env));
            }
```

</details>

dgramop commented:
excited to see these changes land. I took a peek at the way you source context in jni-min and it strikes me as pretty sound.  I was a bit worried about possible compatibility issues between ndk-glue and android_activity based crates - but I don't use ndk-glue (and it seems android_activity is generally better anyhow?)

wuwbobo2021 commented:
Well, I'm not familiar with `ndk-glue`. Applications based on this crate needs an `ANativeActivity_onCreate` cdylib entry too, see <https://docs.rs/ndk-macro/0.3.0/src/ndk_macro/expand.rs.html>. `ndk_glue::init` called in this entry function initializes `ndk_context` just like `android_activity`.

I don't know if `ndk-glue` and `android_activity` can be dependencies of the same application crate, but I guess it's not suggested, see "Middleware Crates" section in <https://crates.io/crates/android-activity>.

kevinmehall commented:
Did I somehow push to your `main` branch instead of mine? Though ` b0b4a54` was never on this branch, so I'm not sure why the force-push message is showing here. In any case, I restored and saved a copy of this branch at https://github.com/kevinmehall/nusb/compare/wuwbobo2021-android.

wuwbobo2021 commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,588 @@
+    let ver_minor: u16 = ver_parser.next().unwrap();
+
+    Ok(DeviceInfo {
+        devinst: Arc::new(env.new_global_ref(dev)?),
```
Sorry for the mismatch. The name is not from the Android API, and it's not a conventional name used in other Android support crate. I will change the name according to your suggestion.

wuwbobo2021 commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,588 @@
+        }
+        self.vendor_id == other.vendor_id
+            && self.product_id == other.product_id
+            && self.id() == other.id()
```
I just did a test and found that `UsbDevice::getDeviceId()` returns integer decimal `1004` for `/dev/bus/usb/001/004`; it became `1005`, `1006`... when plugging over and over again.

Description of `UsbDevice.getDeviceName()`: "Returns the name of the device. In the standard implementation, this is the path of the device file for the device in the usbfs file system."

Despite of the bus number and device address number "contained" in the returned values of `getDeviceId()` and `getDeviceName()`, I may still ignore the possibility of extracting and providing these values since there might be non-standard implementations for the Android API.

kevinmehall replied:
Yeah, WebUSB also doesn't expose bus number and address, so I made limited them to Windows, Mac, Linux in https://github.com/kevinmehall/nusb/pull/158. There is maybe some use case for the bus number on Android laptops or tablets with multiple USB ports (if they're even separate host controllers), but I think it would be fine to skip those fields.

wuwbobo2021 commented:
No, it's definitely not caused by your operation. I gave up merging conflicts online and synchronized my branch forcefully. I should consider my modifications again before reopening the PR. Of course I have an offline backup of my previous attempt.

wuwbobo2021 committed:
Add Android support

kevinmehall commented:
I'd lean towards making the permission request explicit, since it's an extra step that an Android developer will need to think about and figure out where it fits in their app's workflow.

I was hoping we could add a common API to handle permission requests on both Android and WebUSB, but they work differently enough that I don't think that will be possible. On WebUSB, you can't list devices prior to requesting permission. The [permission request API](https://developer.mozilla.org/en-US/docs/Web/API/USB/requestDevice) takes a set of filters, the user is prompted with a list of devices matching those filters (in browser-owned UI), and only the selected and approved device is returned. While on Android, it looks like the filtering and selection of the device is up to the app developer, and the permission is requested for a specific device.

My plan was to add an API that accepts a set of WebUSB-style filters but potentially returns multiple devices. On desktop OSs it would just filter the `list_devices()` list and return immediately since it doesn't need a permission prompt. However, on Android, if the app passed a filter that matches multiple devices, I don't think we'd want to display a permission prompt for each one. The app developer would need to implement their own UI to choose a device between listing devices and requesting permission. Maybe that's an edge case though -- while it's possible to use a hub and multiple devices on a phone, normally there would only be one.

wuwbobo2021 committed:
Correct a wrong Android-specific note

wuwbobo2021 commented:
It seems like the `EndpointReadUntilShortPacket::read` also refuses to return any data until the internal buffer is filled. I'm expecting to receive data shorter than the internal buffer size within the timeout limitation. I ported `android_usbser::CdcSerial` to Linux and tested the `EndpointReadUntilShortPacket` usage on the PC, the same problem is observed.

Should this be a seperate issue? Would you like to provide some example usage(s) for `EndpointReadUntilShortPacket`?

kevinmehall commented:
I think you're misunderstanding what `EndpointReadUntilShortPacket` is supposed to do.

Transfers complete when:
  * the buffer is filled
  * or a packet smaller than the maximum packet size (short packet) is received
  * or an error or cancellation occurs

This is at the xHCI controller level, and applies to all USB transfers in nusb v0.1 and v0.2, libusb, the in-kernel URB APIs backing the kernel serial drivers, etc.

`EndpointRead` and `EndpointReadUntilShortPacket` are built on top of transfers and can't return any data until the transfer completes. The only difference between them is that `EndpointReadUntilShortPacket`  returns EOF after reading the contents of a transfer that completed due to a short packet. This is for protocols where a short packet is used as a terminator or delimiter between messages, probably not useful for serial.

What serial devices have to do is flush data by sending a short or zero length packet to make the transfer complete so the host receives the data immediately. Hardware serial chips do this either with a pin that forces a flush, or a "latency timer" that flushes after a timeout after the last byte. Firmware implementations often have a `flush` API that you must call.

If you use a buffer size larger than the max packet size, note that in the case of a flush at a packet boundary, you must send a zero-length packet to end the transfer.

wuwbobo2021 commented:
<https://github.com/kevinmehall/nusb/pull/150#issuecomment-3326947087> is still my problem... however I shouldn't have made such a mistake if I had ported `android-usbser` to `nusb` v0.2 by myself.

I added logging for VCP data receiving in my test case, not relying on the loopback data, and found that it's a problem of sending data through `CdcSerial`. Short data *can* be received from the VCP device immediately, but after writing it back I must flush the `EndpointWrite` to see the loopback string received at the other side.

I just created another version of `android-usbser` making use of the `nusb` version for this PR: <https://github.com/wuwbobo2021/android-usbser-rs/tree/nusb-with-android>. To make this PR mergeable, I can remove the `nusb::check_startup_intent` function for now.

wuwbobo2021 committed:
Merge branch 'kevinmehall:main' into main

wuwbobo2021 committed:
Adjust the usage of `Notify` in `android.rs`

wuwbobo2021 commented:
Sorry for the delay of making this branch synchronized again.

wuwbobo2021 committed:
Merge branch 'kevinmehall:main' into main

wuwbobo2021 committed:
Update `jni` to 0.22.3; update `jni-min-helper` to 0.4.0

wuwbobo2021 committed:
CI: Test with stable Rust 1.85

wuwbobo2021 committed:
Merge branch 'kevinmehall:main' into main

wuwbobo2021 committed:
android: remove `check_startup_intent`; fix `getParcelableExtra` usage

wuwbobo2021 committed:
Fix problems on new Android versions

kevinmehall commented on src/platform/linux_usbfs/device.rs:
```diff
@@ -832,12 +832,23 @@ impl LinuxEndpoint {
+                    "Access denied while allocating zero-copy buffer of length {len} for endpoint {}, using fallback",
+                    self.inner.address
+                );
+                Ok(Buffer::new(len))
```
The only place this is called already falls through to `Buffer::new` if this returns any `Err`, so it should already be fine. https://github.com/kevinmehall/nusb/blob/0732f03d44edc66c007412845abc908a5df1d885/src/device.rs#L722-L731

wuwbobo2021 replied:
Oh, I made this before fixing another issue. I will revert this unnecessary change.

kevinmehall committed:
android: Docs and other small tweaks

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -399,7 +406,8 @@ impl PermissionRequest {
             Self::on_receive(&inner_weak, &dev_info_2, env, intent)
         })?;
-        receiver.register_for_action(ACTION_USB_PERMISSION)?;
+        let str_perm = android_app_package_name().to_string() + "." + ACTION_USB_PERMISSION;
```
Noticed `const ACTION_USB_PERMISSION: &str = "rust.android_usbser.USB_PERMISSION";` above.

Now that the package name is used, does this still need the hard-coded `rust.android_usbser` prefix? Should it be a function used in both places?

```rs
fn action_usb_permission() -> String {
    format!("{}.USB_PERMISSION", android_app_package_name())
}
```

wuwbobo2021 replied:
I will change it to use such a function, probably preserving the `rust.android_usbser` prefix to avoid conflicting with other components of the application.

kevinmehall reviewed (commented):
Testing this, working great on Android 16. 🎉

wuwbobo2021 committed:
Merge branch 'kevinmehall:main' into main

kevinmehall commented on src/enumeration.rs:
```diff
@@ -283,10 +311,38 @@ impl DeviceInfo {
+    /// `DeviceInfo::has_permission` returns `false`. Status of the `PermissionRequest`
+    /// can be checked on the activity resume event.
+    #[cfg(target_os = "android")]
+    pub fn request_permission(&self) -> Result<Option<crate::PermissionRequest>, Error> {
```
The return type of this API seems confusing.

First the  `Result`: Typically an async (or in this case MaybeFuture) API would return `impl MaybeFuture<Output = Result<_, _>>` rather than `Result<impl MaybeFuture, _>` so users would expect to write `d.request_permission().await?` rather than `d.request_permission()?.await`. One case where an asynchronous error may be better is if the device is disconnected while the prompt is displayed; this currently returns `false` as if the user denied the request, but maybe should return a Disconnected error like it currently checks for and returns only synchronously.

Is the `Option` part necessary to know whether it already has permission or could it just return a `PermissionRequest` that completes instantly with `true` if it already has the permission? The user can call `has_permission` beforehand themselves if they need to know.

Could this instead just return  `impl MaybeFuture<Output = Result<bool, Error>>`? I'm not sure if the public `device_info` and `responsed` (should be `response` if kept) methods are useful enough to justify exposing the `PermissionRequest` type.

wuwbobo2021 replied:
Well, I will try to change this API. Thanks for the suggestion.

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,625 @@
+    is_instance_of = {
+        AndroidContext,
+    }
+}
```
Is this unused?

wuwbobo2021 replied:
Yes, it is unused after removing `check_startup_intent`. I will delete this.

wuwbobo2021 committed:
android: rework of permission API

wuwbobo2021 committed:
Merge branch 'kevinmehall:main' into main

wuwbobo2021 committed:
android: fix potential dead lock in `HotplugWatch`

wuwbobo2021 committed:
android: do permission request in `open_device`

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,640 @@
+            };
+            // Another thread executing `from_device_info` will block here, until the guard
+            // for the current thread is dropped after `LinuxDevice::create_inner`.
+            let _guard = env.lock_obj(usb_man).unwrap();
```
Can you document what race condition this lock is guarding against? You mentioned a while ago that `UsbManager` is shared, but all of its methods seem to just be IPC calls to a service, so it's not clear why that would not be thread safe, or why this is any different than other methods.

wuwbobo2021 replied:
Well, now I'm afraid that this is not adding any protection if the underlying service or kernel is buggy. Removed this lock.

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,640 @@
+    let (ver_major, ver_minor): (u16, u16) = ver_parser(&version).unwrap_or_else(|| {
+        log::warn!("Unable to parse device USB version '{version}'");
+        (0, 0)
+    });
```
In my testing, I have a device that `lsusb` reports as:

```
bcdUSB               2.10
bcdDevice            2.23
```

and nusb on Linux reports

```
device_version: 0x0223,
usb_version: 0x0210,
```

but this code is returning  `usb_version: 0x0217,`. I logged the `dev.get_version(env)?.to_string()` and `getVersion` is returning `"2.23"`.

So

1. `getVersion` is apparently `bcdDevice`, not `bcdUSB` and should go in `device_version` not `usb_version`. There seems to be conflicting information about this online. 
2. It's being parsed incorrectly.

https://android.googlesource.com/platform/frameworks/base/+/refs/heads/main/services/usb/java/com/android/server/usb/descriptors/UsbDeviceDescriptor.java#94 seems to be where Android is formatting that string, and it comes from bcdDevice and uses "normal" binary coded decimal.

Maybe this changed across Android versions?

kevinmehall replied:
I dug into this and it looks like they did change from `bcdUSB` to `bcdDevice` as well as how the BCD is formatted to a string.

Originally `services/core/jni/com_android_server_UsbHostManager.cpp` called `int version = usb_device_get_version(device);` which is [gets it from bcdUSB](https://github.com/aosp-mirror/platform_system_core/blob/a3b721a32242006b59cb12bd62c9133632af3a2d/libusbhost/usbhost.c#L564)

```c
int usb_device_get_version(struct usb_device *device)
{
    struct usb_device_descriptor *desc = (struct usb_device_descriptor *)device->desc;
    return desc->bcdUSB;
}
```

It passed this to `beginUsbDeviceAdded` which formatted

```java
// Create version string in "%.%" format
String versionString = Integer.toString(version >> 8) + "." + (version & 0xFF);
```

Then https://github.com/aosp-mirror/platform_frameworks_base/commit/513160fd807e43c6a8268cbe279d8398b44acb26 moved the descriptor parsing from C++ to Java.

`UsbDeviceDescriptor::mDeviceRelease` was already being parsed from the `bcdDevice` field

```java
private int mDeviceRelease; // 12:2 Device Release number - BCD
```

```java
mProductID = stream.unpackUsbShort();
mDeviceRelease = stream.unpackUsbShort();
mMfgIndex = stream.getByte();
```

and the formatting was changed to use that:

```java
// Create version string in "%.%" format
String versionString =
        Integer.toString(mDeviceRelease >> 8) + "." + (mDeviceRelease & 0xFF);
```

Then a [later change](https://github.com/aosp-mirror/platform_frameworks_base/commit/608ec66d62647f60c3988922fead33fd7e07755e) replaced the formatting with real BCD:


```diff
+    // mDeviceRelease is binary-coded decimal, format DD.DD
+    public String getDeviceReleaseString() {
+        int hundredths = mDeviceRelease & 0xF;
+        int tenths = (mDeviceRelease & 0xF0) >> 4;
+        int ones = (mDeviceRelease & 0xF00) >> 8;
+        int tens = (mDeviceRelease & 0xF000) >> 12;
+        return String.format("%d.%d%d", tens * 10 + ones, tenths, hundredths);
+    }
```

```diff
-        // Create version string in "%.%" format
-        String versionString =
-                Integer.toString(mDeviceRelease >> 8) + "." + (mDeviceRelease & 0xFF);
-        String serialStr = parser.getDescriptorString(mSerialNum);
+        String versionString = getDeviceReleaseString();
+        String serialStr = getSerialString(parser);
```

It looks like both of those changes happened between Android 8 and 9.

Probably only need to support the new one? It looks like there's no way to get `bcdDevice` on Android 8, and no way to get `bcdUSB` on 9+.

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,640 @@
+                PendingIntent::FLAG_MUTABLE
+            };
+            let pending_intent = PendingIntent::get_broadcast(env, context, 0, intent, flags)?;
+            usb_man.request_permission(env, dev_info.jni_global_ref.as_ref(), pending_intent)
```
One edge case with unfortunate behavior is if you have two devices and two threads try to open them simultaneously (presumably same with async). It only shows one permission dialog, and the other request never returns. It may be an Android limitation that it only handles one at a time.

We could consider a workaround on our side like an async mutex or queue to request permissions in sequence (sounds complicated to integrate with MaybeFuture), or even just a static `AtomicBool` tracking that a request is pending to make the second one fail instead of hang indefinitely.

But this is kind of an edge case, so if you can't think of a better way to handle this easily, I don't consider it a blocking issue for this PR. A user can always do their own queue outside of nusb (or open a PR) if they hit this.

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -0,0 +1,640 @@
+        env: &mut Env<'a>,
+        intent: Intent<'a>,
+    ) -> Result<(), jni::errors::Error> {
+        let dev = get_extra_device(env, &intent)?;
```
There are a couple places where you call `get_extra_device` and construct a full `DeviceInfo` just to get the device ID. Performance of this is probably not significant so this may be premature optimization, but it might make sense to have a `get_extra_device_id` that just gets the ID instead of the full `DeviceInfo`.

kevinmehall commented on src/enumeration.rs:
```diff
@@ -256,9 +273,22 @@ impl DeviceInfo {
+        #[cfg(target_os = "android")]
+        {
+            self.get_serial_number()
+        }
```
Especially after the change to put `request_permission` in `open`, I'm thinking it's too magical to have `serial_number` (and thus `Debug` and things) perform a JNI call to re-try getting the serial number after the `DeviceInfo` is created. Every other `DeviceInfo` field on every other platform is plain old data gathered once when it is constructed in `list_devices` or the hotplug event, so the serial number should probably be checked once in `build_device_info`, and if it fails just leave it `None`.

I've been thinking about adding some new APIs that can fetch un-cached descriptors if necessary and that kind of heavier request might fit better there.

kevinmehall reviewed (commented):
I think this is getting very close to ready. I just added a commit with some documentation edits and other minor tweaks, and here are a few more questions from my review and testing today.

wuwbobo2021 committed:
android: add `get_extra_device_id`, remove serial number tricks

wuwbobo2021 committed:
android: make permission request exclusive, fix device version parser

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -522,7 +580,10 @@ pub fn open_device(dev: &DeviceInfo) -> impl MaybeFuture<Output = Result<Arc<Lin
-            Ok(LinuxDevice::create_inner(owned_fd))
+            let linux_dev = LinuxDevice::create_inner(owned_fd).inspect_err(|_| {
+                let _ = conn.close(env);
+            });
```
The `OwnedFd` now owns the file descriptor, and if a `LinuxDevice` is not created to take ownership of it, it will be dropped and close the file descriptor. So I think this will double-close it on error.

wuwbobo2021 replied:
Sorry, I will revert this change.

kevinmehall commented on src/platform/linux_usbfs/android.rs:
```diff
@@ -232,38 +234,67 @@ fn build_device_info(
+                    None
+                }
+            }
+        },
```
Any reason not to catch the exception on all versions?

wuwbobo2021 replied:
Well... my previous thought is that this shouldn't produce an exception under old Android versions. I will change it to simplify the code.

kevinmehall committed:
android: put BroadcastReceiver for permission request in a static Mutex instead of spawning a thread to wait for it

wuwbobo2021 committed:
android: try to make `PermissionRequest` more reliable

wuwbobo2021 commented:
I just rebased my branch on main `v0.2.7` and solved messy problems caused by Github conflict solver. Previous commit history is preserved at <https://github.com/uglyoldbob/nusb/tree/android-prev>. (this also reduced my commit count so that it matches with my actual contribution to `nusb`)

[android-usbser-rs](https://github.com/uglyoldbob/android-usbser-rs) is also updated to work with this branch of `nusb`.
